### PR TITLE
Added basic lstm-ctc script. LSTM and LSTM-CTC recipes for SWBD

### PR DESCRIPTION
**Pull request by vijayaditya**
This PR doesn't yet address #207 .  It has been created for collaboration with other members making similar efforts.

The recipe is still being run, so it will be updated as bugs are found.

To checkout this pull request locally see https://help.github.com/articles/checking-out-pull-requests-locally/

### CODE REVIEW

**Comment by danpovey** [comment]
I'll merge this now and check it later.

File: egs/swbd/s5c/local/nnet3/run_lstm.sh [hunk 1]
```diff
@@ -19,36 +31,43 @@ hidden_dim=1024
 recurrent_projection_dim=256
 non_recurrent_projection_dim=256
 chunk_width=20
-chunk_left_context=20
-clipping_threshold=5.0
+chunk_left_context=40
```
Comment (danpovey) [inline egs/swbd/s5c/local/nnet3/run_lstm.sh]: we're using chunk-left-context=40 now?

Comment (vijayaditya) [inline egs/swbd/s5c/local/nnet3/run_lstm.sh]: Yes this was slightly better.

Vijay

Comment (danpovey) [inline egs/swbd/s5c/local/nnet3/run_lstm.sh]: OK.  It might be worthwhile looking again at increasing the frames-per-eg, then (e.g. to 30?), and maybe reducing the minibatch size or learning rate slightly if you worry about stability (at least, until @xiaohui-zhang  checks in his changes to the max-change code).  It just seems to me that having the left-context much more than the frames-per-eg is a bit wasteful.

Comment (vijayaditya) [inline egs/swbd/s5c/local/nnet3/run_lstm.sh]: This was a temporary thing until the state preservation feature was
implemented.

But I will try out chunk-width of 30 and left context of 20. With momentum
the instabilities were not seen. So I will just try this setting without
modifying chunks-per-minibatch or learning rate.

**Comment by gmkim90** [comment]
Thanks to share the script !  This is exact script what I was looking for. 

About 4months ago, I implemented my own CTC for nnet2, and test it with DNN(non-recurrent)-CTC 
but for almost all utterances, p(blank symbol) >> p(non-blank symbol) during whole utterance. My conjecture is that CTC requires RNN or very deep(so that it can utilize long enough context) DNN to work.

**Comment by danpovey** [comment]
In Google, during decoding they scale blank probabilities down by a factor
of 10 (i.e. multiply by 0.1).  In my scripts, after tuning that scale, I
had to use a less aggressive value of 0.4 (i.e. close to 1/2), but what I
implemented is not vanilla CTC, but an extension that I'm calling CCTC
(described in the link), so I was expecting to have to do less aggressive
scaling of the blank probability.
Dan

**Comment by gmkim90** [comment]
For google's work, you are referring this paper : "Fast and Accurate RNN AM for Speech Recognition"(http://arxiv.org/pdf/1507.06947.pdf), right?

**Comment by danpovey** [comment]
Yes.  It says "In decoding, we only scale the blank label posterior by a
constant scalar decided by cross-validation on a held-out set.".  The paper
doesn't seem to mention it, but I think that constant scalar is either 1/9
or 1/10.  I either heard this from one of the authors, or from some other
paper.
Dan

File: egs/swbd/s5c/local/nnet3/run_lstm.sh (hunk 1 above)
Comment (vijayaditya) [inline egs/swbd/s5c/local/nnet3/run_lstm.sh]: %WER 44.1  VD-LSTM with chunk-width 30 and chunk-left-context 30 (Total training time is 17:04:14)
%WER 43.8  VD-LSTM with chunk-width 20 and chunk-left-context 40 (Total training time is 18:43:35)

Not much benefit in terms of speed and slightly worse results. 

I ensured that the frames-per-iter is the same in both cases by using 
13334 chunks per iteration x 30 chunk-width
20000 chunks per iteration x 20 chunk-width

VD-LSTM : Variable delay LSTM

**Comment by gmkim90** [comment]
I have read Google's recent papers dealing with CCTC and running script of ctc_swbd branch. (download from https://github.com/vijayaditya/kaldi/tree/ctc_swbd)

I have one question : which version of cctc does this code use? I wonder current swbd script use all the newly introduced techniques in KALDI CCTC (of hidden documentation at http://kaldi-asr.org/doc2/ctc.html) uploaded by manager. Specifically, does this code use feature listed below?
 - frame decimation
 - left-context phone unit
 - phone level LM
 - CTC probability normalization
 - LM correction factor at decoding

**Comment by danpovey** [comment]
>
> I have read Google's recent papers dealing with CCTC and running script of
> ctc_swbd branch. (download from
> https://github.com/vijayaditya/kaldi/tree/ctc_swbd)
>
The Google papers were about CTC.  CCTC is an acronym I coined to describe
what I actually implemented, which is a kind of extension, including the
things that you describe below, and as described in
http://kaldi-asr.org/doc2/ctc.htm <http://kaldi-asr.org/doc2/ctc.html>.
All of those things are implemented in the code in the ctc branch.

Bear in mind that it is still being developed- while it's running OK with
regular DNNs, and we haven't got the CTC+LSTM working yet (we're still
tuning the scripts).
Dan

I have one question : which version of cctc does this code use? I wonder

**Comment by gmkim90** [comment]
Thank you for reply.

As far as I understand, ctc branch provide script for wsj (and may use all of the features in homepage documentation) and ctc_swbd provide script for swbd by @vijayaditya. 
I just wonder whether ctc_swbd use same features so that I can directly compare ctc_swbd's script and documentation in homepage. I am checking them by myself now. 

Anyway, thanks! It is really interesting script&code to analyze.
